include/swift/AST/Ownership.h

```diff
@@ -31,21 +33,76 @@ namespace swift {
 enum class ReferenceOwnership : uint8_t {
   /// \brief a strong reference (the default semantics)
   Strong,
+  Default = Strong,
 
-  /// \brief a 'weak' reference
-  Weak,
+#define REF_STORAGE(Name, ...) Name,
+#define REF_STORAGE_RANGE(First, Last) Last_Kind = Last
+#include "swift/AST/ReferenceStorage.def"
+};
+
+enum : unsigned { NumReferenceOwnershipBits =
+  countBitsUsed(static_cast<unsigned>(ReferenceOwnership::Last_Kind)) };
+
+static inline llvm::StringRef keywordOf(ReferenceOwnership ownership) {
+  switch (ownership) {
+  case ReferenceOwnership::Strong:
+    llvm_unreachable("not applicable");
+  case ReferenceOwnership::Weak: return "weak";
+  case ReferenceOwnership::Unowned: return "unowned";
+  case ReferenceOwnership::Unmanaged: return "unowned(unsafe)";
+  }
+  llvm_unreachable("impossible");
+}
 
-  /// \brief an 'unowned' reference
-  Unowned,
+static inline llvm::StringRef manglingOf(ReferenceOwnership ownership) {
+  switch (ownership) {
+  case ReferenceOwnership::Strong:
+    llvm_unreachable("not applicable");
+  case ReferenceOwnership::Weak: return "Xw";
+  case ReferenceOwnership::Unowned: return "Xo";
+  case ReferenceOwnership::Unmanaged: return "Xu";
+  }
+  llvm_unreachable("impossible");
+}
 
-  /// \brief an 'unowned(unsafe)' reference
-  Unmanaged,
+static inline int strengthOf(ReferenceOwnership ownership) {
+  // A reference can be optimized away if outlived by a stronger reference.
+  // NOTES:
+  // 1) Different reference kinds of the same strength are NOT interchangable.
+  // 2) Static references must be last to preserve identity comparisons until
+  //    the last dynamic reference is dead.
+  // 3) Please keep the switch statement ordered to ease code review.
+  switch (ownership) {
+  case ReferenceOwnership::Strong: return 0;
+  case ReferenceOwnership::Unowned: return -1;
+  case ReferenceOwnership::Weak: return -2;
+#define STATIC_REF_STORAGE(Name, ...) \
+  case ReferenceOwnership::Name: return INT_MIN;
+#include "swift/AST/ReferenceStorage.def"
+  }
+  llvm_unreachable("impossible");
+}
```

Comment:
What I want is tristate comparison and returning `int` seemed like the easiest way to help clients do greater-than/less-than/equal-to comparisons.

Reply:
But why not just give them a comparator?  It can return a non-bool if you want to do simultaneous comparison.  But the numbers here are meaningless, and they define away partiality, which I expect will sometimes be the right result (should we really be "strengthening" weak references to unowned or vice-versa?).